app/ydoc-shared/src/ast/tree.ts

```diff
@@ -139,12 +197,12 @@ export abstract class Ast {
     return metadata as FixedMapView<NodeMetadataFields & MetadataFields>
   }
 
-  /** Get metadata of all widgets assigned to this node. */
-  widgetsMetadata(): FixedMapView<Record<string, unknown>> {
+  /** Get metadata of all widgets assigned to this node. WARNING: The returned value must not be mutated. */
+  widgetsMetadata(): ReadonlyMap<string, DeepReadonly<unknown>> {
```

Comment:
On this occasion, add docs for `nodeMetadata` above (with similar warning if needed).

Reply:
Added docs. `nodeMetadata` is safe, we just can't enforce readonly access to `widgetsMetadata` because it doesn't have a know type to apply the `DeepReadonly` transformation to.